[GSoC 2018] LIMS Integration into EHR

Perseus opened the pull request:
I had to create a new branch due to some rebase/merge issues with my earlier one.

What this branch currently contains:

* Base directory structure supporting multiple LIMS implementations
* Guzzle as the cURL library
* Documentation both for the user and the developer

The following features are fully functional in the branch: 

* Login
* Dashboard
* Instrument Management ( Locations, Types and Instruments )
* Manufacturers and Suppliers Management
* Methods
* Laboratory Management ( Lab contacts, Lab departments and Lab Products )
* Analysis Management ( Clients so far, there were some issues with the Senaite API regarding updating data so I'm currently figuring that out with the help of their community )


WIP: 
* ~~Sample Management (Types, Points, Matrices, Conditions)~~
* ~~Calculations (This goes hand in hand with the Methods, I need to create a small parser that is able to calculate based on a string provided by the user )~~
* ~~Analysis ( Categories, Requests, Services, Profiles )~~
* ~~Worksheets ( Templates and the Worksheets themselves )~~
* ~~Data transfer between the LIMS and EHR~~
* ~~Containers (Containers, Container Types)~~
* Ability to create new users (This has to deal directly with Plone and not Senaite, looking into the limitations for this)


Screenshots: 
![image](https://user-images.githubusercontent.com/6716458/42682814-37679e58-86a9-11e8-9128-0936d7d0256b.png)

![image](https://user-images.githubusercontent.com/6716458/42682816-3995f36e-86a9-11e8-9a64-8f3d2767b347.png)

![image](https://user-images.githubusercontent.com/6716458/42682821-3c3b63b0-86a9-11e8-9934-5e2319107935.png)

![image](https://user-images.githubusercontent.com/6716458/42682829-3eb104c4-86a9-11e8-989f-9323faff9b43.png)
The UI is completely responsive and usable on mobiles and desktops alike. I intend to re-do the design for the dashboard to something more pleasing after I've finished the rest of the project.

aethelwulffe commented:
Very very nice.

aethelwulffe commented:
I really hope you can manage a demo of this.  The project is...takes a pretty good lot of upcurve of learning!

aethelwulffe commented:
Any updates this week?

Perseus commented:
Yup, sorry, I'll push a commit asap. Been having some issues with the API

aethelwulffe commented:
"Having issues with the API" is good enough of an update (and *real* 
easy to understand the validity of).  It is just the weekly minimum 
checking thing! ;-)


On 2018-08-01 09:50, Anirudh Singh wrote:
>
> Yup, sorry, I'll push a commit asap. Been having some issues with the API
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/LibreHealthIO/lh-ehr/pull/1204#issuecomment-409581576>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AAhzFyrlP5xK1NquODYqwDPYd5uyoh3Iks5uMbJAgaJpZM4VOitz>.
>

Perseus commented:
What I've added with this commit:

* Sample Management (Types, Points, Matrices, Conditions)
* Calculations (This goes hand in hand with the Methods, I need to create a small parser that is able to calculate based on a string provided by the user )
* Analysis ( Categories, Requests, Services, Profiles )
* Worksheets ( Templates and the Worksheets themselves )
* Data transfer between the LIMS and EHR
* Workflow diagram on how to go about setting it all up and manage procedure orders

How the data integration currently works : 

1. User issues procedure order in the EHR 
2. Procedure orders show up in the LIMS and can be accepted by lab management faculty
3. Once accepted, analysis requests can be created and assigned to those procedure orders
4. Each step of the Analysis Request( sample received, worksheet created, results updated) are reflected in a table I created linking procedures and analysis requests.
5. Once the AR is resolved and a result is produced, results are updated in the procedure_order_results table and also a page/pdf is created for printing with all the results in them.

There are currently some issues that I'm unable to solve due to senaite's API. I've raised the issues on their github and am waiting for resolution to be able to proceed. 
Those can be seen here: 
https://github.com/senaite/senaite.jsonapi/issues/27
https://github.com/senaite/senaite.jsonapi/issues/28 ( I made a temporary workaround for this so this should work )
https://github.com/senaite/senaite.jsonapi/issues/29 ( Due to this issue, unique sample types cannot be created leading to a difficulty creating analysis requests. The code is all there, it will function the moment this issue is fixed )


I've also updated the developer docs to include instructions on updating/maintaining the latest version of senaite, 

![image](https://user-images.githubusercontent.com/6716458/43999705-74d08460-9e2f-11e8-8471-956b48d70e78.png)
Picture of the workflow, can also be found in PDF form in the documentation folder.

WIP: 

Creating new users (analysts) - I'm adding this functionality to senaite's code so will have to create a PR and wait for it to be merged. Till then, I can upload a version of the code that has the function in it. ( need input here )

@aethelwulffe @teryhill 
Sorry for the long gap between commits, quite the learning curve figuring out the API. 
Also, what can I do to remove the other branch commits from this PR? I rebased the branch and it showed no issues but commits from other branches are showing up. Is that how it's supposed to be?

teryhill commented:
@tmccormi can you rebase this and clean up the commit?

aethelwulffe commented on modules/LIMS/senaite/templates/pages/site/worksheets/worksheet.php:
```diff
@@ -71,3 +87,22 @@
+  
+}
+
+</script>
```
Newline!  :)

Perseus commented:
@aethelwulffe I'll get right on it. 

For the user issue, I'm working on integrating a user creator into the LIMS itself. I had a couple interviews the last few days so I wasn't able to finish that feature. 

The sampletype issue is the one that was holding back the creation of unique analysis requests. I did add in the code that should work once that issue is resolved but I can't be sure unless I test it. It should be fixed over the next few days and everything else will work.

Thanks for the kind words 😄
